core/src/Yaya/Strict.hs

```diff
@@ -0,0 +1,159 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE Safe #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Yaya.Strict
+  ( IsStrict,
+    IsNotStrict,
+    Strict,
+    PartialTypeError,
+    unsatisfiable,
+  )
+where
+
+import "base" Data.Bool (Bool (False, True))
+import "base" Data.Either (Either)
+import "base" Data.Functor.Const (Const)
+import "base" Data.Functor.Identity (Identity)
+import "base" Data.List.NonEmpty (NonEmpty)
+import "base" Data.Maybe (Maybe)
+import "base" Data.Type.Bool (Not)
+import "base" Data.Void (Void)
+import "base" GHC.TypeError
+  ( Assert,
+    ErrorMessage (ShowType, Text, (:$$:), (:<>:)),
+    TypeError,
+  )
+import "base" GHC.TypeLits (Symbol)
+import "base" Numeric.Natural (Natural)
+import "comonad" Control.Comonad.Trans.Env (EnvT)
+import "free" Control.Comonad.Cofree (Cofree)
+import "free" Control.Monad.Trans.Free (Free, FreeF, FreeT)
+import qualified "lens" Control.Lens as Lens
+import qualified "strict" Data.Strict as Strict
+import "base" Prelude (error)
+
+-- | This tags types (and type constructors) with whether or not they are
+--   strict.
+--
+--   A type is only strict when all the following conditions hold:
+-- - no referenced type is non-strict
+-- - no non-phantom type parameter is applied to a non-strict type
+-- - all unapplied type parameters are strict (or phantom)
+--
+--  __FIXME__: This needs a third case (so, @`Maybe` `Bool`@) to distinguish
+--            “simple” types that can be used birecursively, because they have
+--             no recursive structure.
+type family Strict (t :: k) :: Bool
```

Comment:
This is the main new thing added, and this needs instances for _many_ cases.

On the plus side, if an instance is missing, the type that’s missing an instance won’t work with anything, so they can be added as needed.